{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d484eec1-982c-4aad-8071-034b8cd73a47",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(data.table) #loading packages\n",
    "library(dplyr)\n",
    "library(tidyr)\n",
    "library(ggplot2)\n",
    "library(VennDiagram)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "39537e3d-920a-400d-8d0b-d6340a14fc16",
   "metadata": {},
   "source": [
    "#### 0.计算每一个subfamily表达的loci比例【CountTable】。"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9e9b41d7-b284-461e-81d3-8e23f6b685a6",
   "metadata": {},
   "outputs": [],
   "source": [
    "setwd(\"/home/xxzhang/workplace/project/CRISPRa/expression/ESC-div/TElocal/\") #完整运行的数据，以后可以直接使用绝对路径；\n",
    "data<-read.table(\"TE.cntTable\",header=TRUE)\n",
    "ExtensionDat<- data %>% separate(gene.TE, c(\"transcript\",\"gene\",\"family\",\"class\"),sep = \":\")\n",
    "FilterDat<-ExtensionDat\n",
    "colnames(FilterDat)[5:13]<-c(\"ESC-1\",\"ESC-2\",\"ESC-3\",\"Neuron-1\",\"Neuron-2\",\n",
    "            \"Neuron-3\",\"Neuron_pro-1\",\"Neuron_pro-2\",\"Neuron_pro-3\")\n",
    "FilterDat_long <- gather(FilterDat, key = \"sample\", value = \"expression\",\n",
    "                    -`transcript`, -`gene`, -`family`, -`class`)\n",
    "CountTable<-FilterDat_long %>% group_by(sample,gene) %>% summarize(notZero = sum(expression!=0),total=sum(!is.na(expression)))\n",
    "CountTable$Percentage<-CountTable$notZero/CountTable$total\n",
    "colnames(CountTable)[2]<-\"TE_subfamily\"\n",
    "head(CountTable)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "011d7971-94c5-418b-a026-84a9c3ecf96d",
   "metadata": {},
   "source": [
    "#### 1.从整体上看，表达的家族的更年轻。"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6cfc7d4c-79b9-4af0-b4e2-392cb1d159e3",
   "metadata": {},
   "outputs": [],
   "source": [
    "#获取亚家族层面的注释信息，包括a.平均长度 b.变异程度 c.家族内部序列的数目\n",
    "setwd(\"/home/xxzhang/workplace/project/CRISPRa/expression/ESC-div/TElocal/DEseq2/\")\n",
    "fa.bed.o<-fread(\"hg38.fa.out\",fill=T,header=T)\n",
    "fa.bed<-fa.bed.o[c(-1,-2),]\n",
    "colnames(fa.bed) <- c(\"SW_score\", \"perc_div\", \"perc_del\", \"perc_ins\", \"query_seq\", \"pos_in_query_begin\", \"pos_in_query_end\", \"pos_in_query_left\", \"strand\", \"TE_subfamily\", \"class_family\", \"pos_in_repeat_begin\", \"pos_in_repeat_end\", \"pos_in_repeat_left\", \"ID\")\n",
    "fa.bed$length <- as.numeric(fa.bed$pos_in_query_end) - as.numeric(fa.bed$pos_in_query_begin)\n",
    "mean_length <- aggregate(fa.bed$length, by=list(fa.bed$TE_subfamily, fa.bed$class_family), FUN=mean)\n",
    "colnames(mean_length) <- c(\"TE_subfamily\", \"info\", \"avg_length\")\n",
    "mean_per_div <- aggregate(as.numeric(fa.bed$perc_div), by=list(fa.bed$TE_subfamily, fa.bed$class_family), FUN=mean)\n",
    "colnames(mean_per_div) <- c(\"TE_subfamily\", \"info\", \"avg_perc_div\")\n",
    "te_freq <- as.data.frame(table(fa.bed$TE_subfamily))\n",
    "te_freq <- merge(te_freq, fa.bed[,c(\"TE_subfamily\", \"class_family\")], by.x=\"Var1\", by.y=\"TE_subfamily\")\n",
    "colnames(te_freq) <- c(\"TE_subfamily\", \"Freq\", \"info\")\n",
    "te_freq.2<-distinct(te_freq)\n",
    "mergeDat<-merge(mean_length,mean_per_div,by=\"TE_subfamily\")\n",
    "mergeDat.2<-merge(mergeDat,te_freq.2,by=\"TE_subfamily\")\n",
    "finalDat<-mergeDat.2[,c(1,2,3,5,6)]\n",
    "write.csv(finalDat,\"finalDat.csv\") #这个结果就是我们想要的\n",
    "############################################################################\n",
    "annotation<-read.csv(\"./DEseq2/finalDat.csv\",row.names=1) #这个数据集来源于网页，关于那个还是应该更加细致的了解；\n",
    "annotationDat<-merge(annotation,CountTable,by=\"TE_subfamily\")[,c(-2,-3,-4,-5,-7,-8)] #与表达矩阵进行各种层面的整合；\n",
    "head(annotationDat)\n",
    "annotationDat.2 <- aggregate(annotationDat$Percentage, by=list(annotationDat$TE_subfamily,annotationDat$avg_perc_div),mean)\n",
    "colnames(annotationDat.2)<-c(\"TE_subfamily\",\"avg_perc_div\",\"mean_perc\") \n",
    "#原来还对length的规律进行了查看；目前还是仅仅关注avg_perc_div(可以表征时间的一个变量)\n",
    "options(repr.plot.width =4, repr.plot.height =4)\n",
    "p<-ggplot(annotationDat.2,aes(x=avg_perc_div,y=mean_perc))+\n",
    "    geom_point(size=1,shape=15)+\n",
    "    geom_smooth(method=lm)+\n",
    "    theme_minimal()\n",
    "pdf('cor.pdf',width = 4,height = 4)\n",
    "p\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1a683554-98e9-472b-b8fb-8a1ca94e1284",
   "metadata": {},
   "source": [
    "#### 2.从整体上看，长度更长的转录本，更倾向于表达。"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "69d4f32c-2af7-4c98-b287-715ea82bda3d",
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf<-read.table(\"/home/xxzhang/data/Genome_reference/GTF/TE_gtf/L1PA2.locInd.locations\") #这是仅仅考虑了一个家族，但是实际上可以考虑更多家族\n",
    "gtff<-gtf %>% separate(V2, c(\"chromsome\",\"start\",\"stop\",\"strand\"),sep = \"[-|:]\")\n",
    "colnames(gtff)[1]<-\"transcript\"\n",
    "gtff$start<-as.numeric(gtff$start)\n",
    "gtff$stop<-as.numeric(gtff$stop)\n",
    "gtff$length<-gtff$stop-gtff$start\n",
    "LengthDat<-merge(gtff,FilterDat) #与上述表达矩阵进行整合\n",
    "LengthDat_long <- gather(LengthDat, key = \"sample\", value = \"expression\",\n",
    "                    -`transcript`,-`chromsome`, -`start`, -`stop`, -`strand` ,-`length`, -`gene`, -`family`, -`class`)\n",
    "LengthDat_long$length<-as.numeric(LengthDat_long$length)\n",
    "LengthDat_long$Length_label[LengthDat_long$length>6000] <-\">6k\"\n",
    "LengthDat_long$Length_label[LengthDat_long$length<=6000 & LengthDat_long$length>4000 ] <-\"4k-6k\"\n",
    "LengthDat_long$Length_label[LengthDat_long$length<=4000 & LengthDat_long$length>2000 ] <-\"2k-4k\"\n",
    "LengthDat_long$Length_label[LengthDat_long$length<= 2000] <-\"<=2k\"\n",
    "LengthDat_long$Length_label<-factor(LengthDat_long$Length_label,levels=c(\"<=2k\",\"2k-4k\",\"4k-6k\",\">6k\"))\n",
    "options(repr.plot.width =10, repr.plot.height =4)\n",
    "p <- ggplot(LengthDat_long,\n",
    "            aes(sample,log10(expression+1),fill = Length_label)) +\n",
    "       #geom_violin(trim=FALSE)+\n",
    "       #geom_boxplot(width=0.1)+\n",
    "       geom_boxplot(color=\"lightgrey\")+\n",
    "       theme_classic()+\n",
    "       #coord_cartesian(ylim = c(0, 10))+\n",
    "       # scale_fill_brewer(palette=\"Blues\")\n",
    "           scale_fill_brewer(palette=\"BuGn\")\n",
    "pdf('L1PA2-Length2.pdf',width = 8,height = 4)\n",
    "p\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7ccc524d-336e-4b1a-be88-3eed79e5726a",
   "metadata": {},
   "source": [
    "#### 3.从整体上看，结构更完整的转录本，更倾向于表达。"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eb466a14-8ab9-4618-8514-d045e4905385",
   "metadata": {},
   "outputs": [],
   "source": [
    "setwd(\"/home/xxzhang/workplace/software/ORFFinder/\")\n",
    "data<-read.table(\"query-hg38_bedtools_L1.chr1-Y.notAnnotation.txt\")\n",
    "colnames(data)<-c(\"query.acc.ver\",\"subject.acc.ver\", \"%identity\", \n",
    "                  \"alignment.length\",\n",
    "                  \"mismatches\",\"gap.opens\",\"q.start\",\n",
    "                  \"q.end\",\"s.start\",\"s. end\",\"evalue\",\"bit.score\")\n",
    "data_ext<-data %>% separate(query.acc.ver, c(\"lcl\",\"subfamily\",\"null\",\"chromsome\",\"start-stop\",\"strand\",\"orf.start-orf.end\"),sep = \"[:|(]\")\n",
    "data_ext2<-data_ext %>% separate(`start-stop`, c(\"start\",\"stop\"),sep = \"[-]\")\n",
    "data_ext3<-data_ext2 %>% separate(`orf.start-orf.end`, c(\"orf.start\",\"orf.end\"),sep = \"[-]\")\n",
    "data_ext3[which(data_ext3$strand==\"+)\"),]$strand<-\"+\"\n",
    "data_ext3[which(data_ext3$strand==\"-)\"),]$strand<-\"-\"\n",
    "data_ext3[grepl(\"ORF2\",data_ext3$subject.acc.ver),\"ORF\"]<-\"ORF2\"\n",
    "data_ext3[grepl(\"ORF1\",data_ext3$subject.acc.ver),\"ORF\"]<-\"ORF1\"\n",
    "data_ext3$Length<-as.numeric(data_ext3$stop)-as.numeric(data_ext3$start) #数据整理\n",
    "data_ext4<-data_ext3[,c(2,4,5,6,7,22,8,9,11,12,13,14,15,16,17,18,21)] #从这边开始，重新做\n",
    "data_ext4[,\"label\"]<-\"null\"\n",
    "data_ext4<-data_ext4[order(data_ext4$subfamily,data_ext4$chromsome,data_ext4$start,\n",
    "                           data_ext4$stop,data_ext4$strand,data_ext4$Length,data_ext4$ORF),] #加上这一行重新开始做；对数据进行排序，否则对下游的结果影响很大\n",
    "for (i in 1:(dim(data_ext4)[1]-1)) {\n",
    "if(sum(data_ext4[i,c(1:6,17)]==data_ext4[i+1,c(1:6,17)])==7){\n",
    "    a=c(data_ext4[i,7],data_ext4[i,8],data_ext4[i+1,7],data_ext4[i+1,8])\n",
    "    data_ext4[i,]$label<-\"1rd\"\n",
    "    data_ext4[i+1,]$label<-\"1rd\"\n",
    "    b=as.data.frame(c(data_ext4[i,1:6],min(a),max(a),rep(NA,8),data_ext4[i,17],\"2rd\"))\n",
    "    colnames(b)<-colnames(data_ext4)\n",
    "    data_ext4<-rbind(data_ext4,b)\n",
    "    i = i+1\n",
    "}\n",
    "    }\n",
    "data_ext5<-data_ext4[!grepl(\"1rd\", data_ext4$label),] #删除1rd的结果\n",
    "lines<-c() \n",
    "for (j in 1:(dim(data_ext5)[1]-1)) {\n",
    "if(sum(data_ext5[j,c(1:6,17,18)]==data_ext5[j+1,c(1:6,17,18)])==8){\n",
    "    lines<-append(lines,j) #有一点小聪明\n",
    "    lines<-append(lines,j+1)\n",
    "    a=c(data_ext5[j,7],data_ext5[j,8],data_ext5[j+1,7],data_ext5[j+1,8])\n",
    "    #data_ext5[i,]$label<-\"1rd\"\n",
    "    #data_ext5[i+1,]$label<-\"1rd\"\n",
    "    b=as.data.frame(c(data_ext5[j,1:6],min(a),max(a),rep(NA,8),data_ext5[j,17],\"3rd\"))\n",
    "    colnames(b)<-colnames(data_ext5)\n",
    "    data_ext5<-rbind(data_ext5,b)\n",
    "    j = j+1\n",
    "}\n",
    "    }\n",
    "data_ext6<-data_ext5[-lines,]\n",
    "lines2<-c()\n",
    "for (m in 1:(dim(data_ext6)[1]-1)) {\n",
    "if(sum(data_ext6[m,c(1:6,17)]==data_ext6[m+1,c(1:6,17)])==7){\n",
    "    a=c(data_ext6[m,7],data_ext6[m,8],data_ext6[m+1,7],data_ext6[m+1,8])\n",
    "    lines2<-append(lines2,m) #有一点小聪明\n",
    "    lines2<-append(lines2,m+1)\n",
    "    #data_ext6[i,]$label<-\"1rd\"\n",
    "    #data_ext6[i+1,]$label<-\"1rd\"\n",
    "    b=as.data.frame(c(data_ext6[m,1:6],min(a),max(a),rep(NA,8),data_ext6[m,17],\"4rd\"))\n",
    "    colnames(b)<-colnames(data_ext6)\n",
    "    data_ext6<-rbind(data_ext6,b)\n",
    "    m = i+1\n",
    "}\n",
    "    }\n",
    "data_ext7<-data_ext6[-lines2,]\n",
    "data_ext8<-data_ext7[order(data_ext7$subfamily,data_ext7$chromsome,data_ext7$start,\n",
    "                           data_ext7$stop,data_ext7$strand,data_ext7$Length,\n",
    "                          data_ext7$ORF),]\n",
    "data_ext8[,\"class\"]<-\"null\"\n",
    "for (m in 1:(dim(data_ext8)[1]-1)) {\n",
    "#print(m)\n",
    "if(sum(data_ext8[m,c(1:6)]==data_ext8[m+1,c(1:6)])==6){\n",
    "    #if(data_ext8[m,17]==\"ORF1\"& data_ext8[m+1,17]==\"ORF2\"){\n",
    "        data_ext8[m,19]<-\"ORF1-ORF2\"\n",
    "        data_ext8[m+1,19]<-\"ORF1-ORF2\" #注意区分\n",
    "        #print(m)\n",
    "        m = m+2 #即使程序运行完了加2，都无法改变1，2，3，4遍历的事实（想一下办法怎么改变这个局面）\n",
    "        #print(m)\n",
    "    #}\n",
    "}  \n",
    "}\n",
    "    #print(m)\n",
    "for (m in 1:(dim(data_ext8)[1]-1)) {\n",
    "#print(m)\n",
    "if(data_ext8[m,19]!=\"ORF1-ORF2\"){\n",
    "     if(data_ext8[m,17]==\"ORF1\"){\n",
    "        data_ext8[m,19]<-\"ORF1\"\n",
    "    }else{\n",
    "        data_ext8[m,19]<-\"ORF2\"\n",
    "}  \n",
    "}\n",
    "    }\n",
    "data_ext8[12175,19]<-\"ORF2\"\n",
    "data_ext8$label<-paste(paste(paste(paste(data_ext8$chromsome,data_ext8$start,sep=\":\"),data_ext8$stop,sep=\"-\"),data_ext8$strand,sep=\"(\"),\"\",sep=\")\")\n",
    "data_ext8$label2<-paste(paste(paste(data_ext8$chromsome,data_ext8$start,sep=\":\"),data_ext8$stop,sep=\"-\"),data_ext8$strand,sep=\":\")\n",
    "write.table(data_ext8,\"data_ext8_ordered.final.txt\",row.names=F,quote=F,sep=\"\\t\") #这个数据就是最终的对L1结构的注释信息\n",
    "################################################################################################\n",
    "structure<-read.table(\"/home/xxzhang/workplace/software/ORFFinder/data_ext8_ordered.final.txt\",header=T)\n",
    "structure2<-unique(structure[,c(20,1,19)])\n",
    "colnames(structure2)[1]<-c(\"location\")\n",
    "gtf<-read.table(\"/home/xxzhang/data/Genome_reference/GTF/TE_gtf/hg38_rmsk_TE.gtf.locInd.locations\",header=T)\n",
    "colnames(gtf)<-c(\"transcriptID\",\"location\")\n",
    "annotationDat<-merge(gtf,structure2,by=\"location\",all=TRUE)\n",
    "annotationDat[is.na(annotationDat$class),]$class<-\"not-enough-length\"\n",
    "FilterDat2<-FilterDat[FilterDat$family==\"L1\",] #表达的结果中仅仅考虑L1 #FilterDat为上述的表达矩阵\n",
    "StructureDat<-merge(annotationDat,FilterDat2,by=\"transcriptID\")\n",
    "CountTable<-StructureDat%>% count(gene,class.x,name = 'count')\n",
    "library(RColorBrewer)\n",
    "colourCount <-  length(unique(CountTable$class.x))\n",
    "p<-ggplot(CountTable, aes(gene,log10(count+1),fill = class.x))+\n",
    "    geom_col(position = 'stack', width = 0.7) +\n",
    "    labs(x = '', y = 'Relative Abundance(%)') +\n",
    "    theme(axis.text = element_text(size = 12), axis.title = element_text(size = 13)) +\n",
    "    theme(legend.text = element_text(size = 11))+\n",
    "    theme_classic()+\n",
    "    scale_fill_manual(values = colorRampPalette(brewer.pal(8, \"Set2\"))(colourCount))+\n",
    "    theme(legend.position = \"bottom\" ,legend.box = \"horizontal\")+\n",
    "    coord_flip()\n",
    "pdf('L1-count.pdf',width = 7,height = 20)\n",
    "p\n",
    "dev.off()\n",
    "CountTable2<-CountTable%>% filter(gene%in%c(\"L1HS\",\"L1PA2\",\"L1PA3\",\"L1PA4\",\"L1PA5\",\"L1PA6\",\"L1PA7\",\"L1PA8\"))\n",
    "library(RColorBrewer)\n",
    "colourCount <-  length(unique(CountTable2$class.x))\n",
    "options(repr.plot.width =6, repr.plot.height =3)\n",
    "p<-ggplot(CountTable2, aes(gene,log10(count+1),fill = class.x))+\n",
    "    geom_col(position = 'stack', width = 0.7) +\n",
    "    labs(x = '', y = 'log10(count+1)') +\n",
    "    theme(axis.text = element_text(size = 12), axis.title = element_text(size = 13)) +\n",
    "    theme(legend.text = element_text(size = 11))+\n",
    "    theme_classic()+\n",
    "    scale_fill_manual(values = colorRampPalette(brewer.pal(8, \"Set2\"))(colourCount))+\n",
    "    theme(legend.position = \"bottom\" ,legend.box = \"horizontal\")+\n",
    "    coord_flip()\n",
    "pdf('L1-count-subset.pdf',width = 6,height = 3)\n",
    "p\n",
    "dev.off()\n",
    "#############################################################################################\n",
    "StructureDat_long <- gather(StructureDat, key = \"sample\", value = \"expression\",\n",
    "                    -`transcriptID`,-`location`, -`subfamily`, -`class.x`, -`gene` ,-`family`, -`class.y`)\n",
    "options(repr.plot.width =12, repr.plot.height =4)\n",
    "p <- ggplot(StructureDat_long,\n",
    "            aes(sample,log10(expression+1),fill = class.x)) +\n",
    "       geom_boxplot(color=\"lightgrey\")+\n",
    "       theme_classic()+\n",
    "           scale_fill_brewer(palette=\"BuGn\")\n",
    "pdf('L1-structure.pdf',width = 10,height = 4)\n",
    "p\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "89de260d-b0d2-4ee8-ae48-f92a6cc21672",
   "metadata": {},
   "source": [
    "#### 4.表达的loci和gRNA靶向的loci之间的关系。"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "42144840-0e67-474a-a19d-9eaa95e43388",
   "metadata": {},
   "outputs": [],
   "source": [
    "FilterDat_noZero_L1<-FilterDat_long %>% filter(expression!=0,gene%in%c(\"L1HS\",\"L1PA2\",\"L1PA3\",\"L1PA4\"))\n",
    "FilterDat_noZero_L1$group[FilterDat_noZero_L1$sample%in%c(\"ESC-1\",\"ESC-2\",\"ESC-3\")] <-\"ESC\"\n",
    "FilterDat_noZero_L1$group[FilterDat_noZero_L1$sample%in%c(\"Neuron_pro-1\",\"Neuron_pro-2\",\"Neuron_pro-3\")] <-\"Neuron_pro\"\n",
    "FilterDat_noZero_L1$group[FilterDat_noZero_L1$sample%in%c(\"Neuron-1\",\"Neuron-2\",\"Neuron-3\")] <-\"Neuron\"\n",
    "repFilterDat_noZero_L1<-FilterDat_noZero_L1 %>% count(group,transcript,gene, sort = TRUE)\n",
    "Savedfiles_repFilterDat_noZero_L1<-repFilterDat_noZero_L1 %>% filter(n==3)\n",
    "gtf<-read.table(\"/home/xxzhang/data/Genome_reference/GTF/TE_gtf/L1PA1-4.locInd.locations\")\n",
    "gtff<-gtf %>% separate(V2, c(\"chromsome\",\"start\",\"stop\",\"strand\"),sep = \"[-|:]\")\n",
    "colnames(gtff)[1]<-\"transcript\"\n",
    "gtff$start<-as.numeric(gtff$start)\n",
    "gtff$stop<-as.numeric(gtff$stop)\n",
    "gtff$length<-gtff$stop-gtff$start\n",
    "Merge_Savedfiles_repFilterDat_noZero_L1<-merge(gtff,Savedfiles_repFilterDat_noZero_L1)\n",
    "Merge_Savedfiles_repFilterDat_noZero_L1$strand[Merge_Savedfiles_repFilterDat_noZero_L1$strand%in%c(\"\")] <-\"-\"\n",
    "Merge_Savedfiles_repFilterDat_noZero_L1$strand[Merge_Savedfiles_repFilterDat_noZero_L1$strand%in%c(\"-\")] <-\"(-)\"\n",
    "Merge_Savedfiles_repFilterDat_noZero_L1$strand[Merge_Savedfiles_repFilterDat_noZero_L1$strand%in%c(\"+\")] <-\"(+)\"\n",
    "Merge_Savedfiles_repFilterDat_noZero_L1$label<-paste(paste(Merge_Savedfiles_repFilterDat_noZero_L1$chromsome,paste(Merge_Savedfiles_repFilterDat_noZero_L1$start, Merge_Savedfiles_repFilterDat_noZero_L1$stop, sep=\"-\"),sep=\":\"),\n",
    "     Merge_Savedfiles_repFilterDat_noZero_L1$strand,sep=\"\")\n",
    "blast_Result<-read.table(\"/home/xxzhang/data/TE/Result/blast.txt\")\n",
    "blast_Result2<-blast_Result %>% separate(V2, c(\"gene\",\"label\"),sep = \"::\")\n",
    "blast_Merge_Savedfiles_repFilterDat_noZero_L1<-merge(blast_Result2,Merge_Savedfiles_repFilterDat_noZero_L1,by=\"label\",all=TRUE) \n",
    "#根据label如chrY:10855136-10855667(-)的重合程度来判断是否target的是同一个loci\n",
    "gRNA1<-blast_Result2[blast_Result2$V1==\"TGGGAGATATACCTAATGCTAGATGACACA\",]$label\n",
    "gRNA2<-blast_Result2[blast_Result2$V1==\"CCGGCTTAAGAAACGGCGCACCACGAGACT\",]$label\n",
    "gRNA4<-blast_Result2[blast_Result2$V1==\"TATCCGCTGTTCTGCAGCCACCGCTGCTGA\",]$label\n",
    "gRNA5<-blast_Result2[blast_Result2$V1==\"GAAACTTCCAGAGGAACGATCAGGCAGCAA\",]$label\n",
    "Vennplot_gRNA<-function(gRNA,gRNA_name,celltype){\n",
    "    L1HS<-Merge_Savedfiles_repFilterDat_noZero_L1[Merge_Savedfiles_repFilterDat_noZero_L1$gene==\"L1HS\"&Merge_Savedfiles_repFilterDat_noZero_L1$group==celltype,]$label\n",
    "    L1PA2<-Merge_Savedfiles_repFilterDat_noZero_L1[Merge_Savedfiles_repFilterDat_noZero_L1$gene==\"L1PA2\"&Merge_Savedfiles_repFilterDat_noZero_L1$group==celltype,]$label\n",
    "    L1PA3<-Merge_Savedfiles_repFilterDat_noZero_L1[Merge_Savedfiles_repFilterDat_noZero_L1$gene==\"L1PA3\"&Merge_Savedfiles_repFilterDat_noZero_L1$group==celltype,]$label\n",
    "    L1PA4<-Merge_Savedfiles_repFilterDat_noZero_L1[Merge_Savedfiles_repFilterDat_noZero_L1$gene==\"L1PA4\"&Merge_Savedfiles_repFilterDat_noZero_L1$group==celltype,]$label\n",
    "    \n",
    "    \n",
    "    venn.diagram(x=list(gRNA,L1HS,L1PA2,L1PA3,L1PA4),\n",
    "             scaled = F, # 根据比例显示大小\n",
    "             alpha= 1, #透明度\n",
    "             lwd=1,lty=1,col=c(\"#b0c8e9\",\"#ffbc79\",\"#99de8b\",\"#227ab5\", \"#966bbd\"), #圆圈线条粗细、形状、颜色；1 实线, 2 虚线, blank无线条\n",
    "             label.col ='black' , # 数字颜色abel.col=c('#FFFFCC','#CCFFFF',......)根据不同颜色显示数值颜色\n",
    "             cex = 0.8, # 数字大小\n",
    "             fill=c(\"#b0c8e9\",\"#ffbc79\",\"#99de8b\",\"#227ab5\", \"#966bbd\"), # 填充色 配色https://www.58pic.com/\n",
    "             category.names = c(gRNA_name, \"L1HS\",\"L1PA2\",\"L1PA3\",\"L1PA4\") , #标签名\n",
    "             cat.dist = c(0.2, 0.2, 0.2, 0.2, 0.2), # 标签距离圆圈的远近\n",
    "             cat.pos = c(0, -10, 240, 120, 20), # 标签相对于圆圈的角度cat.pos = c(-10, 10, 135)\n",
    "             cat.cex = 1, #标签字体大小\n",
    "             fontfamily=3,\n",
    "             cat.fontfamily=3,\n",
    "             cat.col=c('black','black',\"black\",\"black\", \"black\"),   #cat.col=c('#FFFFCC','#CCFFFF',.....)根据相应颜色改变标签颜色\n",
    "             cat.default.pos = \"outer\",  # 标签位置, outer内;text 外\n",
    "             output=TRUE,\n",
    "             filename=paste(gRNA_name,celltype,\"png\",sep=\".\"),# 文件保存\n",
    "             imagetype=\"png\",  # 类型（tiff png svg）\n",
    "             resolution = 400,  # 分辨率\n",
    "             compression = \"lzw\",# 压缩算法\n",
    "             main=celltype,\n",
    "             main.cex = 2, main.fontfamily = 3,\n",
    " \n",
    ")\n",
    "    \n",
    "}\n",
    "\n",
    "Vennplot_gRNA(gRNA1,\"gRNA1\",\"ESC\") #gRNA1\n",
    "Vennplot_gRNA(gRNA1,\"gRNA1\",\"Neuron_pro\")\n",
    "Vennplot_gRNA(gRNA1,\"gRNA1\",\"Neuron\")\n",
    "Vennplot_gRNA(gRNA2,\"gRNA2\",\"ESC\") #gRNA2\n",
    "Vennplot_gRNA(gRNA2,\"gRNA2\",\"Neuron_pro\")\n",
    "Vennplot_gRNA(gRNA2,\"gRNA2\",\"Neuron\")\n",
    "Vennplot_gRNA(gRNA4,\"gRNA4\",\"ESC\") #gRNA3\n",
    "Vennplot_gRNA(gRNA4,\"gRNA4\",\"Neuron_pro\")\n",
    "Vennplot_gRNA(gRNA4,\"gRNA4\",\"Neuron\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "15a4c07a-ea0d-43de-9767-7b817cb7ec32",
   "metadata": {},
   "source": [
    "#### 5.统计差异的loci的数目，及其所贡献的subfamily【感觉没啥意义】。"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a0a33078-729f-461d-9c5b-bb0b39e01461",
   "metadata": {},
   "outputs": [],
   "source": [
    "setwd(\"/home/xxzhang/workplace/project/CRISPRa/expression/ESC-div/TElocal/DEseq2/\")\n",
    "data<-read.table(\"/home/xxzhang/workplace/project/CRISPRa/expression/ESC-div/TElocal/DEseq2/ESC_sigdiff_gene_TE.txt\")\n",
    "TE.esc<-cbind(row.names(data[!grepl(\"ENSG*\", row.names(data)),]),data[!grepl(\"ENSG*\", row.names(data)),])  \n",
    "TE.esc.row.names<- TE.esc %>% separate(TE, c(\"transcript\",\"gene\",\"family\",\"class\"),sep = \":\")\n",
    "TE.neuron.count<-as.data.frame(table(TE.esc.row.names[TE.esc.row.names$log2FoldChange<0,\"gene\"]))\n",
    "TE.esc.count<-as.data.frame(table(TE.esc.row.names[TE.esc.row.names$log2FoldChange>0,\"gene\"]))\n",
    "mergeDat<-merge(TE.esc.count,TE.neuron.count,by=\"Var1\",all=TRUE)\n",
    "colnames(mergeDat)<-c(\"subfamily\",\"ESC\",\"neuron/neuronalprogentor\")\n",
    "mergeDat_long <- gather(mergeDat, key = \"group\", value = \"lDEoci\",\n",
    "                    -`subfamily`)\n",
    "options(repr.plot.width =6, repr.plot.height =100)\n",
    "library(RColorBrewer)\n",
    "colourCount <-  length(unique(mergeDat_long$group))\n",
    "p<-ggplot(mergeDat_long, aes(subfamily,log10(lDEoci+1),fill = group))+\n",
    "    geom_col(position = 'dodge', width = 0.7) +\n",
    "    labs(x = 'subfamily', y = 'log10(DEloci)') +\n",
    "    theme(axis.text = element_text(size = 4), axis.title = element_text(size = 5)) +\n",
    "    theme(legend.text = element_text(size = 11))+\n",
    "    theme_classic()+\n",
    "    scale_fill_manual(values = colorRampPalette(brewer.pal(8, \"Set2\"))(colourCount))+\n",
    "    theme(legend.position = \"bottom\" ,legend.box = \"horizontal\")+\n",
    "    coord_flip()\n",
    "pdf('TE-DEloci.pdf',width = 6,height = 100)\n",
    "p\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "37f5202d-cfd0-4695-8eec-7c4f204ca018",
   "metadata": {},
   "source": [
    "> 感觉我现在的技术还很十分不高超。只会简单的计数，连统计都不太会。要去文献中多学习一些。"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.2.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
